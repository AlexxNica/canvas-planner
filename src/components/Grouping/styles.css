/* Variables are defined in ./theme.js */

.groupingList,
.itemList {
  list-style-type: none;
  margin: 0;
  padding: 0;
  border-color: var(--defaultIconColor);
  color: var(--defaultIconColor);
}

.grouping {
  font-family: var(--fontFamily);
  margin-top: var(--groupingTopMargin);
  border-color: var(--defaultIconColor);
}

.groupingHero {
  position: relative;
  background-repeat: no-repeat;
  background-position: center center;
  background-size: cover;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  color: var(--heroLinkColor);
  font-size: var(--heroLinkFontSize);
  font-weight: var(--heroLinkFontWeight);
  letter-spacing: var(--heroLinkLetterSpacing);
  outline: none;
  padding: var(--heroLinkPadding);
  text-decoration: none;

  /* this can become min-height once we drop IE11 support */
  height: var(--groupingHeroMinHeight);

  .groupingName {
    text-decoration: var(--heroLinkTextDecoration);
  }

  &:focus,
  &:hover {
    text-decoration: none;

    .groupingName {
      text-decoration: var(--heroLinkTextDecorationHover);
    }
  }
}

.groupingOverlay {
  background-color: var(--defaultIconColor);
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
}

.groupingName {
  position: relative;
  z-index: 1;
  flex: 1;
  box-sizing: border-box;
  text-align: center;
  padding: var(--heroLinkNamePadding);
  background-color: var(--heroLinkNameBackground);
  text-transform: var(--heroLinkNameTextTransform);
}

[data-media-medium] {
  .grouping {
    display: flex;
  }

  .groupingHero {
    flex: 0 0 var(--groupingHeroWidth);
    height: auto;
  }

  .groupingHero,
  .groupingOverlay {
    border-bottom-left-radius: var(--heroBorderRadius);
    border-top-left-radius: var(--heroBorderRadius);
  }

  .itemList {
    flex: 1;
    border-top: var(--borderTopWidthTabletUp) solid;
  }
}
