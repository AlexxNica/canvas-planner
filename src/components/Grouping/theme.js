/* Global variables (colors, typography, spacing, etc.) are defined in lib/themes */

export default function generator ({ borders, colors, media, spacing, typography }) {
  return {
    fontFamily: typography.fontFamily,
    lineHeight: typography.lineHeightCondensed,
    margin: `${spacing.medium} 0 0 0`,

    groupColor: colors.brand,

    borderTopWidthTabletUp: borders.widthSmall,

    heroMinHeight: '7rem',
    heroWidth: '12rem',
    heroWidthLarge: '14rem',
    heroPadding: '0 0.0625rem',
    heroColor: colors.brand,
    heroBorderRadius: borders.radiusMedium,

    overlayOpacity: 0.75,

    titleFontSize: typography.fontSizeXSmall,
    titleFontWeight: typography.fontWeightBold,
    titleLetterSpacing: '0.0625rem',
    titleBackground: colors.white,
    titleTextTransform: 'uppercase',
    titlePadding: `${spacing.xxSmall} ${spacing.xSmall}`,
    titleTextDecoration: 'none',
    titleTextDecorationHover: 'underline',
    titleColor: colors.brand,
    activityIndicatorPadding: spacing.small,
    activityIndicatorWidth: spacing.small,
    ...media
  };
}

generator['canvas-a11y'] = generator['modern-a11y'] = function ({ colors }) {
  return {
    heroTextDecoration: 'underline',
    heroTextDecorationHover: 'none',
    titleColor: colors.licorice,
  };
};

generator.canvas = function (variables) {
  return {
    groupColor: variables["ic-brand-primary"],
    titleColor: variables["ic-brand-primary"]
  };
};
