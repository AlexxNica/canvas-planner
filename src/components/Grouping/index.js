import React, { Component } from 'react';
import themeable from 'instructure-ui/lib/themeable';
import { arrayOf, string, object } from 'prop-types';
import styles from './styles.css';
import theme from './theme.js';
import PlannerItem from '../PlannerItem';
import moment from 'moment-timezone';

class Grouping extends Component {
  static propTypes = {
    items: arrayOf(object).isRequired,
    courseInfo: object.isRequired,
    timeZone: string.isRequired
  }
  render () {
    return (
      <ol className={styles.groupingList}>
        <li
          className={styles.grouping}
          style={{borderColor: this.props.courseInfo.color}}>
          <a href="#"
            className={styles.groupingHero}
            style={{backgroundImage: `url(${this.props.courseInfo.image_url})`}}>
            <span
              className={styles.groupingOverlay}
              style={{backgroundColor: this.props.courseInfo.color}}></span>
            <span className={styles.groupingName}>
              {this.props.items[0].context.title}
            </span>
          </a>
          <ol
            className={styles.itemList}
            style={{borderColor: this.props.courseInfo.color}}>
            {
              this.props.items.map((item) => {
                return (
                  <PlannerItem
                    key={item.id}
                    color={item.context.color}
                    completed
                    id={item.id}
                    date={moment.tz(item.date, this.props.timeZone).toDate()}
                    associated_item={item.type}
                    title={item.title}
                    points={item.points}
                    toggleCompletion={() => console.log('send me back to canvas')}
                  />
                );
              })
            }
          </ol>
        </li>
      </ol>
    );
  }
}

export default themeable(theme, styles)(Grouping);
