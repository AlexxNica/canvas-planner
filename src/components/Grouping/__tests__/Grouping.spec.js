import React from 'react';
import { shallow, mount } from 'enzyme';
import Grouping from '../index';

const getDefaultProps = () => ({
  items: [{
    id: "5",
    title: 'San Juan',
    date: '2017-04-25T05:06:07-08:00',
    context: {
      url: 'example.com',
      color: "#5678",
      id: 256
    }
  }, {
    id: "6",
    date: '2017-04-25T05:06:07-08:00',
    title: 'Roll for the Galaxy',
    context: {
      color: "#5678",
      id: 256
    }
  }],
  timeZone: "America/Denver",
  color: "#5678",
  id: 256,
  url: 'example.com',
  title: 'Board Games',
  updateTodo: () => {}
});

it('renders the base component with required props', () => {
  const wrapper = shallow(
    <Grouping {...getDefaultProps()} />
  );
  expect(wrapper).toMatchSnapshot();
});

it('grouping contains link pointing to course url', () => {
  const props = getDefaultProps();
  const wrapper = shallow(
    <Grouping {...props} />
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders to do items correctly', () => {
  const props = {
    items: [{
      id: "700",
      title: 'To Do 700',
      date: '2017-06-16T05:06:07-06:00',
      context: null,
    }],
    timeZone: "America/Denver",
    color: null,
    id: null,
    url: null,
    title: null,
    updateTodo: () => {}
  };
  const wrapper = shallow(
    <Grouping {...props} />
  );
  expect(wrapper).toMatchSnapshot();
});

it('does not render completed items by default', () => {
  const props = getDefaultProps();
  props.items[0].completed = true;
  const wrapper = shallow(
    <Grouping {...props} />
  );

  expect(wrapper.find('PlannerItem')).toHaveLength(1);
});

it('renders a CompletedItemsFacade when completed items are present by default', () => {
  const props = getDefaultProps();
  props.items[0].completed = true;

  const wrapper = shallow(
    <Grouping {...props} />
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders completed items when the facade is clicked', () => {
  const props = getDefaultProps();
  props.items[0].completed = true;

  const wrapper = mount(
    <Grouping {...props} />
  );

  wrapper.instance().handleFacadeClick();
  expect(wrapper.find('PlannerItem')).toHaveLength(2);
});

it('renders completed items when they have the show property', () => {
  const props = getDefaultProps();
  props.items[0].show = true;
  props.items[0].completed = true;

  const wrapper = shallow(
    <Grouping {...props} />
  );

  expect(wrapper.find('PlannerItem')).toHaveLength(2);
});

it('does not render a CompletedItemsFacade when showCompletedItems state is true', () => {
  const props = getDefaultProps();
  props.items[0].completed = true;

  const wrapper = shallow(
    <Grouping {...props} />
  );

  wrapper.setState({ showCompletedItems: true });
  expect(wrapper.find('CompletedItemsFacade')).toHaveLength(0);
});

it('renders an activity notification when there are things in the past with status badges', () => {
  const props = getDefaultProps();
  props.items[0].status = ["graded"];
  props.isInPast = true;

  const wrapper = shallow(
    <Grouping {...props} />
  );

  expect(wrapper.find('Badge')).toHaveLength(1);
});

it('does not render an activity badge when things in the past have no status', () => {
  const props = getDefaultProps();
  props.items[0].status = [];
  props.isInPast = true;
  const wrapper = shallow(
    <Grouping {...props} />
  );
  expect(wrapper.find('Badge')).toHaveLength(0);
});

it('invokes the takeFocusRef (if passed) on a focusable element', () => {
  const mockTakeFocus = jest.fn();
  mount(<Grouping {...getDefaultProps()} takeFocusRef={mockTakeFocus} />);
  expect(mockTakeFocus).toHaveBeenCalledWith(expect.anything());
});

describe('handleFacadeClick', () => {
  it('sets focus to the groupingLink when called', () => {
    const wrapper = mount(
      <Grouping {...getDefaultProps()} />
    );
    wrapper.instance().handleFacadeClick();
    expect(document.activeElement).toBe(wrapper.instance().groupingLink);
  });

  it('calls preventDefault on an event if given one', () => {
    const wrapper = mount(
      <Grouping {...getDefaultProps()} />
    );
    const fakeEvent = {
      preventDefault: jest.fn()
    };
    wrapper.instance().handleFacadeClick(fakeEvent);
    expect(fakeEvent.preventDefault).toHaveBeenCalled();
  });
});

describe('toggleCompletion', () => {
  it('binds the toggleCompletion method to item', () => {
    const mock = jest.fn();
    const props = getDefaultProps();
    const wrapper = mount(
      <Grouping
        {...props}
        toggleCompletion={mock}
      />
    );
    wrapper.find('input').first().simulate('change');
    expect(mock).toHaveBeenCalledWith(props.items[0]);
  });
});
