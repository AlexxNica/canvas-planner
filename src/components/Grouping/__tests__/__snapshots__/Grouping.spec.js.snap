// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grouping contains link pointing to course url 1`] = `
<div>
  <div />
  <a
    href="example.com"
    style={
      Object {
        "backgroundImage": "url()",
      }
    }
  >
    <span
      className=""
      style={
        Object {
          "backgroundColor": "#5678",
        }
      }
    />
    <span>
      Board Games
    </span>
  </a>
  <ol
    style={
      Object {
        "borderColor": "#5678",
      }
    }
  >
    <li>
      <PlannerItem
        badges={Array []}
        color="#5678"
        context={
          Object {
            "color": "#5678",
            "id": 256,
            "url": "example.com",
          }
        }
        courseName="Board Games"
        date={"2017-04-25T13:06:07.000Z"}
        id="5"
        theme={
          Object {
            "iconColor": "#5678",
          }
        }
        title="San Juan"
        toggleCompletion={[Function]}
        updateTodo={[Function]}
      />
    </li>
    <li>
      <PlannerItem
        badges={Array []}
        color="#5678"
        context={
          Object {
            "color": "#5678",
            "id": 256,
          }
        }
        courseName="Board Games"
        date={"2017-04-25T13:06:07.000Z"}
        id="6"
        theme={
          Object {
            "iconColor": "#5678",
          }
        }
        title="Roll for the Galaxy"
        toggleCompletion={[Function]}
        updateTodo={[Function]}
      />
    </li>
  </ol>
</div>
`;

exports[`renders a CompletedItemsFacade when completed items are present by default 1`] = `
<div>
  <div />
  <a
    href="example.com"
    style={
      Object {
        "backgroundImage": "url()",
      }
    }
  >
    <span
      className=""
      style={
        Object {
          "backgroundColor": "#5678",
        }
      }
    />
    <span>
      Board Games
    </span>
  </a>
  <ol
    style={
      Object {
        "borderColor": "#5678",
      }
    }
  >
    <li>
      <PlannerItem
        badges={Array []}
        color="#5678"
        context={
          Object {
            "color": "#5678",
            "id": 256,
          }
        }
        courseName="Board Games"
        date={"2017-04-25T13:06:07.000Z"}
        id="6"
        theme={
          Object {
            "iconColor": "#5678",
          }
        }
        title="Roll for the Galaxy"
        toggleCompletion={[Function]}
        updateTodo={[Function]}
      />
    </li>
    <li>
      <CompletedItemsFacade
        badges={Array []}
        itemCount={1}
        onClick={[Function]}
      />
    </li>
  </ol>
</div>
`;

exports[`renders the base component with required props 1`] = `
<div>
  <div />
  <a
    href="example.com"
    style={
      Object {
        "backgroundImage": "url()",
      }
    }
  >
    <span
      className=""
      style={
        Object {
          "backgroundColor": "#5678",
        }
      }
    />
    <span>
      Board Games
    </span>
  </a>
  <ol
    style={
      Object {
        "borderColor": "#5678",
      }
    }
  >
    <li>
      <PlannerItem
        badges={Array []}
        color="#5678"
        context={
          Object {
            "color": "#5678",
            "id": 256,
            "url": "example.com",
          }
        }
        courseName="Board Games"
        date={"2017-04-25T13:06:07.000Z"}
        id="5"
        theme={
          Object {
            "iconColor": "#5678",
          }
        }
        title="San Juan"
        toggleCompletion={[Function]}
        updateTodo={[Function]}
      />
    </li>
    <li>
      <PlannerItem
        badges={Array []}
        color="#5678"
        context={
          Object {
            "color": "#5678",
            "id": 256,
          }
        }
        courseName="Board Games"
        date={"2017-04-25T13:06:07.000Z"}
        id="6"
        theme={
          Object {
            "iconColor": "#5678",
          }
        }
        title="Roll for the Galaxy"
        toggleCompletion={[Function]}
        updateTodo={[Function]}
      />
    </li>
  </ol>
</div>
`;

exports[`renders to do items correctly 1`] = `
<div>
  <div />
  <a
    href="#"
    style={
      Object {
        "backgroundImage": "url()",
      }
    }
  >
    <span
      className=""
      style={
        Object {
          "backgroundColor": null,
        }
      }
    />
    <span>
      To Do
    </span>
  </a>
  <ol
    style={
      Object {
        "borderColor": null,
      }
    }
  >
    <li>
      <PlannerItem
        badges={Array []}
        color={null}
        context={Object {}}
        courseName={null}
        date={"2017-06-16T11:06:07.000Z"}
        id="700"
        theme={
          Object {
            "iconColor": null,
          }
        }
        title="To Do 700"
        toggleCompletion={[Function]}
        updateTodo={[Function]}
      />
    </li>
  </ol>
</div>
`;
